### PR TITLE
[EasyCore] Implement ABN validator in EasyCore package

### PR DESCRIPTION
**Pull request by BOB41K1987**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |

### CODE REVIEW

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    /**
+     * @var int[]
+     */
+    private const WEIGHTS = [10, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19];
+
+    /**
+     * Validates an Australian Business Number.
+     *
+     * @throws \EoneoPay\Utils\Exceptions\UnexpectedTypeException
```
Comment (natepage) [inline packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php:37]: Please let's make sure we move away from this `EoneoPay\Utils` package as it will be deprecated soon. And I think that we should use the `UnexpectedValueException` from Symfony if possible, otherwise let's just create our own exception within easy-core for now.

Comment (egor-dev) [reply packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php:37]: agree

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        if ($value === null || $value === '') {
+            return;
+        }
+
+        if (\is_string($value) === false && (\is_object($value) && method_exists($value, '__toString')) === false) {
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        $abn = (string) $value;
```
Comment (egor-dev) [inline packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php:53]: Let's remove a space after the casting.

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+        if (\strlen($abn) !== self::ABN_LENGTH) {
+            $this->context->buildViolation($constraint->message)
+                ->setCode(Abn::INVALID_LENGTH_ERROR)
+                ->addViolation();
+
+            return;
+        }
+
+        if (\ctype_digit($abn) === false || (int)$abn[0] === 0) {
```
Comment (egor-dev) [inline packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php:63]: Why did you add a second check? I think this check does not match an error code we set here: 0 is not a invalid character.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+        return [
+            'class without __toString' => [new stdClass(), 'Expected argument of type "string", "stdClass" given'],
+            'integer' => [53004085616, 'Expected argument of type "string", "integer" given'],
+        ];
+    }
+
+    /**
+     * @return mixed[]
+     */
+    public function provideValidABN(): array
```
Comment (egor-dev) [inline packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php:49]: please, use `Abn` instead of `ABN`. Here and below.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    public function provideUnexpectedValues(): array
+    {
+        return [
+            'class without __toString' => [new stdClass(), 'Expected argument of type "string", "stdClass" given'],
+            'integer' => [53004085616, 'Expected argument of type "string", "integer" given'],
+        ];
+    }
+
+    /**
+     * @return mixed[]
```
Comment (egor-dev) [inline packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php:47]: `string[][]`?

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+use Symfony\Component\Validator\Violation\ConstraintViolationBuilderInterface;
+
+/**
+ * @covers \EonX\EasyCore\Bridge\Symfony\Validator\Constraints\AbnValidator
+ */
+final class AbnValidatorTest extends AbstractSymfonyTestCase
+{
+    /**
+     * @return mixed[]
```
Comment (egor-dev) [inline packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php:24]: `string[][]`?

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        $this->expectNotToPerformAssertions();
+    }
+
+    /**
+     * @throws \EoneoPay\Utils\Exceptions\UnexpectedTypeException
+     */
+    public function testValidateThrowsUnexpectedTypeException(): void
+    {
+        $validator = new AbnValidator();
+        $constraint = new NotNull();
```
Comment (egor-dev) [inline packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php:147]: Let's use `stdClass` here.

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php (hunk 3 above)
Comment (BOB41K1987) [reply packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php:63]: First of all, first character cannot be 0 according to documentation. Second, when we make `-1` thing below, the first character becomes `-`, not the `-1` as you could expect. That causes an error when we try to multiply digits by weight, because minus sign `-` is not a digit.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php (hunk 7 above)
Comment (BOB41K1987) [reply packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php:147]: We will get an error, `AbnValidator expected instance of Type, got stdClass`, something like this.

Comment (egor-dev) [reply packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php:147]: Got it. Let's create anonymous class extending Constraint then. It's better not to tie with another constraint in the test.

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php (hunk 3 above)
Comment (egor-dev) [reply packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php:63]: Ok, let's introduce `Abn::LEADING_ZERO_ERROR` then to be precise and clear. And move this check to a dedicated if statement.

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php (hunk 1 above)
Comment (BOB41K1987) [reply packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php]: Done.

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php (hunk 2 above)
Comment (BOB41K1987) [reply packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php]: Done.

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php (hunk 3 above)
Comment (BOB41K1987) [reply packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php]: Done.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php (hunk 4 above)
Comment (BOB41K1987) [reply packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php]: Done.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php (hunk 5 above)
Comment (BOB41K1987) [reply packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php]: Done.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php (hunk 6 above)
Comment (BOB41K1987) [reply packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php]: Done.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php (hunk 7 above)
Comment (BOB41K1987) [reply packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php]: Done.

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    private const MODULUS = 89;
+
+    /**
+     * @var int[]
+     */
+    private const WEIGHTS = [10, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19];
+
+    /**
+     * Validates an Australian Business Number.
```
Comment (roman-eonx) [inline packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php:35]: Let's get rid of this useless comment.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        return [
+            'class without __toString' => [new stdClass(), 'Expected argument of type "string", "stdClass" given'],
+            'integer' => [53004085616, 'Expected argument of type "string", "integer" given'],
+        ];
+    }
+
+    /**
+     * @return string[][]
+     */
+    public function provideValidAbn(): array
```
Comment (roman-eonx) [inline packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php:48]: `provideValidAbnValues` to be consistent.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        return [
+            'Empty string' => [''],
+            'Null value' => [null],
+        ];
+    }
+
+    /**
+     * @return string[][]
+     */
+    public function provideInvalidAbn(): array
```
Comment (roman-eonx) [inline packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php:72]: `provideInvalidAbnValues` to be consistent.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+
+        $this->expectNotToPerformAssertions();
+    }
+
+    /**
+     * @dataProvider provideValidEmptyValues
+     *
+     * @param mixed $abn
+     */
+    public function testValidateSucceedsWithEmptyValue($abn): void
```
Comment (roman-eonx) [inline packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php:120]: This method completely repeats `testValidateSucceedsWithValidAbn`.

File: packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php (hunk 8 above)
Comment (BOB41K1987) [reply packages/EasyCore/src/Bridge/Symfony/Validator/Constraints/AbnValidator.php]: Done.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php (hunk 9 above)
Comment (BOB41K1987) [reply packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php]: Done.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php (hunk 10 above)
Comment (BOB41K1987) [reply packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php]: Done.

File: packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php (hunk 11 above)
Comment (BOB41K1987) [reply packages/EasyCore/tests/Bridge/Symfony/Valiator/Constraints/AbnValidatorTest.php]: Done.
